Sources/JExtractSwift/Swift2JavaTranslator+JavaBindingsPrinting.swift

```diff
@@ -124,16 +128,155 @@ extension Swift2JavaTranslator {
+  ///     static MemorySegment toUpcallStub(Function fi, Arena arena) {
+  ///       return Linker.nativeLinker().upcallStub(HANDLE.bindTo(fi), DESC, arena);
+  ///     }
+  ///   }
+  ///   ```
+  func printFunctionPointerParameterDescriptorClass(
+    _ printer: inout CodePrinter,
+    _ name: String,
+    _ cType: CType
+  ) {
+    guard case .pointer(.function(let cResultType, let cParameterTypes, variadic: false)) = cType else {
+      preconditionFailure("must be a C function pointer type")
+    }
+
+    let cParams = cParameterTypes.enumerated().map { i, ty in
+      CParameter(name: "_\(i)", type: ty)
+    }
+    let paramDecls = cParams.map({"\($0.type.javaType) \($0.name!)"})
+
+    printer.printBraceBlock(
+      """
+      /**
+       * {snippet lang=c :
+       * \(cType)
+       * }
+       */
+      private static class \(name)
+      """
+    ) { printer in
+      printer.print(
+        """
+        public interface Function {
+          \(cResultType.javaType) apply(\(paramDecls.joined(separator: ", ")));
+        }
+        """
+      )
+      printFunctionDescriptorDefinition(&printer, cResultType, cParams)
+      printer.print(
+        """
+        private static final MethodHandle HANDLE = SwiftKit.upcallHandle(Function.class, "apply", DESC);
+        private static MemorySegment toUpcallStub(Function fi, Arena arena) {
+          return Linker.nativeLinker().upcallStub(HANDLE.bindTo(fi), DESC, arena);
+        }
+        """
+      )
+    }
+  }
+
+  /// Print the helper type container for a user-facing Java API.
+  ///
+  /// * User-facing function interfaces.
+  func printJavaBindingWrapperHelperClass(
+    _ printer: inout CodePrinter,
+    _ decl: ImportedFunc
+  ) {
+    let translated = self.translatedDecl(for: decl)!
+    let bindingDescriptorName = self.thunkNameRegistry.functionThunkName(decl: decl)
+    if translated.functionTypes.isEmpty {
+      return
+    }
+
+    printer.printBraceBlock(
+      """
+      public static class \(translated.name)
+      """
+    ) { printer in
+      for functionType in translated.functionTypes {
+        printJavaBindingWrapperFunctionTypeHelper(&printer, functionType, bindingDescriptorName)
+      }
+    }
+  }
+
+  /// Print "wrapper" function interface representing a Swift closure type.
+  func printJavaBindingWrapperFunctionTypeHelper(
+    _ printer: inout CodePrinter,
+    _ functionType: TranslatedFunctionType,
+    _ bindingDescriptorName: String
+  ) {
+    let cdeclDescritor = "\(bindingDescriptorName).$\(functionType.name)"
+    if functionType.isTrivial {
+      // If the usser-facing function interface is C-compatible, just extend the
+      // lowered function pointer parameter interface.
+      printer.print(
+        """
+        public interface \(functionType.name) extends \(cdeclDescritor).Function {
+          default MemorySegment toUpcallStub(Arena arena) {
```

Comment:
Maybe it's better to make a static func outside the interface, so it won't leak `toUpcallStub` method?
```java
private static func $toUpcallStub(\(functionType.name) fi, Arena arena) { ... }
```